jerry-debugger/src/lib/parsed-source.ts

```diff
@@ -0,0 +1,178 @@
+// Copyright JS Foundation and other contributors, http://js.foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { PROTOCOL } from './jrs-protocol-constants';
+import { cesu8ToString, concatUint8Arrays } from './string-utils';
+import { JerryDebuggerClient } from './debugger-client';
+import { ParsedFunction } from './breakpoint';
+import { EventEmitter } from 'events';
+
+export interface ParserStackFrame {
+  isFunc: boolean;
+  line: number;
+  column: number;
+  name: string;
+  source: string;
+  sourceName?: string;
+  lines: Array<number>;
+  offsets: Array<number>;
+  byteCodeCP?: number;
+  firstBreakpointLine?: number;
+  firstBreakpointOffset?: number;
+}
+
+export interface FunctionMap {
+  [cp: string]: ParsedFunction;
+}
+
+export class ParsedSource extends EventEmitter {
+  private debugClient: JerryDebuggerClient;
+  private stack: Array<ParserStackFrame>;
+  private source: string = '';
+  private sourceData?: Uint8Array;
+  private sourceName?: string;
+  private sourceNameData?: Uint8Array;
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
+  constructor(debugClient: JerryDebuggerClient) {
```

Comment:
Couple high level comments:

I'd try to avoid node-isms in the parser / protocol code. Relying on `EventEmitter` (a Node API) will make it trickier to make this usable in a browser environment (i.e. the HTML client).
AFAICT there's only one interested party in the `scriptParsed` events, so this mechanism might as well be replaced with the delegate pattern.

Something along these lines:

```

export interface ScriptParsedResult {
  scriptId: number;
  name: string;
  endLine: number;
  endColumn: number,
  hasSourceURL: boolean;
}

export interface ParsedSourceDelegate {
  onScriptParsed(result: ScriptParsedResult): void;
}

export class ParsedSource {
 ...
  // the client would be passed here still, but by having a smaller interface (`ParsedSourceDelegate`) it will
  // make unit testing this simpler because we don't have to mock the whole client, just the delegate methods.
 constructor(public delegate: ParsedSourceDelegate) {
     ...
 }

 ..
 // instead of this.emit('...', obj), we'll do this.delegate.onScriptParsed(obj);

}

```